### DISCUSSION TITLE
Improve frontend scripting capability

### DISCUSSION THREAD

**Discussion by cverdes** [Budibase/budibase#3131] [Ideas]
I have developed a couple of applications in Miscrosoft PowerApp Canvas and I highlight its power of scripting between components and creating global application variables and local screen variables into the frontend.

It uses the [Power Fx language](https://docs.microsoft.com/en-us/power-platform/power-fx/overview) which is basic and simple but it allows much more flexibility in the frontend than currently Budibase.

PowerApps' event-oriented scripting functionality reminds me a lot of first versions of Visual Basic, which, with little and simple coding, allowed non-programmers to create interesting applications.

I think that boosting frontend scripting would be very useful for some advanced users and would give a lot of flexibility and power to the platform.

[Appgyver](https://www.appgyver.com/) bought by SAP is another low code tool that enhances frontend side scripting. Appgyver implements the scripting feature with a visual language as [Visual Scripting of godot](https://docs.godotengine.org/en/latest/tutorials/scripting/visual_script/getting_started.html) or [Scripting visual of Unity3D](https://unity.com/es/products/unity-visual-scripting). 

Perhaps a framework as [Rete.js](https://rete.js.org/#/#lang=en&tosearch=Rete.js%0A) may be integrated into the budibase builder. It would be an honor to collaborate in implementing some of these functionalities.

Regards

**Comment by mjashanks** [comment]
Hi Cesar.

Thanks for your input - it's good to see the comparison with other platforms.

This week, we are releasing some scripting capabilities using JavaScript. This may be used in 2 places:
1. Bindings - when you need to display a value inside a component. Currently, we use the handlebars syntax - `{{ Form.Fields.Forename }}`. Now, we can optionally use javascript, which will enable more complex scripting. I believe that this would cover a lot of the functionality of "Power FX".
2. Transformers. When a query is returned from a REST API for Custom SQL query, Javascript may be used to modify the structure of the data that's returned, in whatever way javascript allows!

We will be adding to this over time. It'd be great if you could give the new functionality a try when released and give us your feedback. It'd be great to hear your thoughts, given that you have experienced other platforms.

The release should be out at the end of this week.

**Reply by samg666** [reply]
Glad to hear that budibase will have javascript support. May I know when will the javascript supported will released?

**Reply by mjashanks** [reply]
This week

**Reply by cverdes** [reply]
@mjashanks  said "_It'd be great if you could give the new functionality a try when released and give us your feedback_"... Yes please! The honor is mine! I want to contribute to the project by testing and improving the documentation as much as I can, under your direction. Count on my interest and help.

Regarding Power FX I found that [the documentation is on github](https://github.com/microsoft/Power-Fx). As the documentation says "_Microsoft Power Fx is a low-code general purpose programming language based on spreadsheet-like formulas_".

I consider the idea of using a language similar to Excel formulas to develop a low code frontend is very smart, since many of the users want to replace their spreadsheets with a low-code application.

If you evaluate that incorporating Power FX into Budibase designer is feasible, or at least if you consider that it justifies do a proof of concept, I could start working on a PowerFX to javascript transpiler and see what comes up.

An advantage of Power FX on the frontend is that you would have the leverage of Microsoft documentation and question-and-answer forums.

What is your opinion regarding the future according to the roadmap already defined for budibase?

Regards

**Reply by mjashanks** [reply]
> Count on my interest and help

Great - thank you!

On Power FX. I can totally see the advantage of offering this as an option - it would really help users transition from power apps to Budibase.

However, given that we are only just at the beginning of our scripting journey, we'll not be looking at it anytime soon. We will aim to flesh out the Javascript offering - there are many places that we can go with it.

It would then be a huge decision down the line to add another scripting option, but Power FX is a great shout!

**Reply by cverdes** [reply]
Glad to hear your vision @mjashanks about the Budibase scripting journey.

I am very interested in collaborating on the Budibase project. I think I can be useful in testing and documentation. Please, count on me as a resource when and where you think it's necessary.

Meanwhile I am going to work, slowly but surely, on seeing how to make Power FX transcript to Javascript, very attentive to Microsoft's open source project.

Willing to help, I offer my free time at your disposal.

**Comment by cverdes** [comment]
Good news! 21 hours ago, Microsoft included the source code of Power FX on github!

![Screenshot from 2021-11-02 19-00-28](https://user-images.githubusercontent.com/6440942/139958019-f1b092f6-3246-4b81-a704-ad9eb42f8eb1.png)

It is developed in C#, but I suppose that, on that basis, an implementation in Typescript should not be very complex.
Let's work!
I'll share my progress in this post.

**Reply by ryzam** [reply]
HI @cverdes any update on your finding/research on power-fx? My development language background also from .Net C#.
